nucypher/network/server.py

```diff
@@ -84,12 +85,24 @@ def make_rest_app(
         log=Logger("http-application-layer")
         ) -> Tuple:
 
-    forgetful_node_storage = ForgetfulNodeStorage(federated_only=this_node.federated_only)
+    # A trampoline function for the real REST app,
+    # to ensure that a reference to the node and the datastore object is not held by the app closure.
+    # One would think that it's enough to only remove a reference to the node,
+    # but `rest_app` somehow holds a reference to itself, Uroboros-like,
+    # and will hold the datastore reference if it is created there.
 
     from nucypher.datastore import datastore
 
     log.info("Starting datastore {}".format(db_filepath))
     datastore = datastore.Datastore(db_filepath)
+    rest_app = _make_rest_app(weakref.proxy(datastore), weakref.proxy(this_node), serving_domains, log)
```

Comment:
To be fair, @jMyles complained when I introduced the change that passed a reference to `this_node` to the rest app. I didn't like how bloated was getting this method call, with >10 attributes from this node and growing, so I thought that maybe it was simpler to pass the node directly. In hindsight, this was contributing to the circular references problem.